src/Futhark/Pass/ExplicitAllocations.hs

```diff
@@ -641,6 +716,64 @@ allocInExp e = mapExpM alloc e
                                                handle op
                          }
 
+addResCtxInIfBody :: (Allocable fromlore tolore, Allocator tolore (AllocM fromlore tolore)) =>
+                     [ExtType] -> Body tolore -> [Maybe Space] ->
+                     [Maybe (Int, ExtIxFun, M.Map Int (PrimExp VName))] ->
+                     AllocM fromlore tolore (Body tolore, [BodyReturns])
+addResCtxInIfBody ifrets (Body _ bnds res) spaces substs = do
+  let num_vals = length ifrets
+      (ctx_res, val_res) = splitFromEnd num_vals res
+      ext_ini = S.size $ shapeContext ifrets
+  ((res', bodyrets'), all_body_stms) <- collectStms $ do
+    mapM_ addStm bnds
+    (val_res', mem_ctx_res, bodyrets, _) <-
+      foldM (\(res_acc, ctx_acc, br_acc, k) (ifr, r, mbixfsub, sp) ->
```

Comment:
Give this giant fold function a local binding.

Reply:
Done!